java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,284 @@
+            )
+        );
+      }
+      this.minPeerTagLength = minPeerTagLength;
+      this.peerKey = peerKey;
+    }
+
+    /**
+     * Initializes user's private key.
+     *
+     * @param userPrivateKey user's private key.
+     * @return current GlomeBuilder object.
+     */
+    public GlomeBuilder setPrivateKey(byte[] userPrivateKey) {
+      this.userPrivateKey = userPrivateKey;
+      return this;
+    }
+
+    /**
+     * Initializes user's private key (if it hadn't been initialized before), user's public key and
+     * each party's MAC key.
+     */
+    public Glome build() throws InvalidKeyException {
+      if (userPrivateKey == null) {
+        this.userPrivateKey = X25519.generatePrivateKey();
+      }
+      this.userPublicKey = X25519.publicFromPrivate(this.userPrivateKey);
+      initMacKeys();
+
+      return new Glome(this);
+    }
+
+    /**
+     * Calculates user's and peer's MAC keys.
+     *
+     * @throws InvalidKeyException if any used key is invalid.
+     */
+    private void initMacKeys() throws InvalidKeyException {
+      byte[] sharedSecret = X25519.computeSharedSecret(userPrivateKey, peerKey);
+      initMacKey(userMacKey, sharedSecret, peerKey, userPublicKey);
+      initMacKey(peerMacKey, sharedSecret, userPublicKey, peerKey);
+    }
+
+    /**
+     * Calculates MAC key from given shared secret, receiver's and sender's public keys.
+     *
+     * @param mac MAC object to be initialized.
+     * @param sharedSecret some shared secret.
+     * @param receiverPublicKey receiver's public key.
+     * @param senderPublicKey sender's public key.
+     */
+    private void initMacKey(Mac mac, byte[] sharedSecret, byte[] receiverPublicKey,
+        byte[] senderPublicKey) throws InvalidKeyException {
+      byte[] key = generateMacMsg(sharedSecret, receiverPublicKey, senderPublicKey);
+
+      SecretKeySpec keySpec = new SecretKeySpec(key, "HmacSHA256");
+      mac.init(keySpec);
+    }
+
+    /**
+     * Generates a message for MAC by concatenating shared secret and two keys.
+     *
+     * @param sharedSecret some shard secret.
+     * @param receiverPublicKey receiver's public key.
+     * @param senderPublicKey sender's public key.
+     * @return generated MAC message.
+     */
+    private byte[] generateMacMsg(byte[] sharedSecret, byte[] receiverPublicKey,
+        byte[] senderPublicKey) {
+      return ByteBuffer
+          .allocate(sharedSecret.length + receiverPublicKey.length + senderPublicKey.length)
+          .put(sharedSecret)
+          .put(receiverPublicKey)
+          .put(senderPublicKey)
+          .array();
+    }
+
+  }
+
+  private Glome(GlomeBuilder builder) {
+    this.userPublicKey = builder.userPublicKey;
+    this.userPrivateKey = builder.userPrivateKey;
+    this.peerKey = builder.peerKey;
+    this.minPeerTagLength = builder.minPeerTagLength;
+    this.userMacKey = builder.userMacKey;
+    this.peerMacKey = builder.peerMacKey;
+  }
+
+  public byte[] peerKey() {
```

Comment:
Can these be mutated? If so, maybe we should return copies of the arrays in these accessors?

Reply:
It can't as the keys fields have `private` modifiers and there are only getters for them.

Reply:
But the returned byte array can be mutated by whoever receives it, no? I.e. it's a reference to the same object and modifications to it will be mirrored to this object as it's the same memory location.

Reply:
Yes, you are right. Fixed.